File: packages/settings/android/src/main/java/io/capkit/settings/SettingsImpl.kt
package io.capkit.settings

import android.content.Context
import android.content.Intent
import io.capkit.settings.config.SettingsConfig
import io.capkit.settings.error.SettingsError
import io.capkit.settings.logger.SettingsLogger
import io.capkit.settings.utils.SettingsUtils

/**
 Native Android implementation for the Settings plugin.

 Responsibilities:
 - Perform platform logic
 - Throw typed SettingsError values on failure

 Forbidden:
 - Accessing PluginCall
 - Referencing Capacitor APIs
 */
class SettingsImpl(
  private val context: Context,
) {
  /**
   * Cached plugin configuration.
   * Injected once during plugin initialization.
   */
  private lateinit var config: SettingsConfig

  /**
   * Applies plugin configuration.
   *
   * This method MUST be called exactly once from the plugin's load() method.
   * It translates static configuration into runtime behavior
   * (e.g. enabling verbose logging).
   */
  fun updateConfig(newConfig: SettingsConfig) {
    this.config = newConfig
    SettingsLogger.verbose = newConfig.verboseLogging
    SettingsLogger.debug(
      "Configuration applied. Verbose logging:",
      newConfig.verboseLogging.toString(),
    )
  }

  /**
   * Opens an Android system settings screen.
   *
   * @throws SettingsError if the operation fails
   */
  fun open(option: String): Intent {
    val intent =
      SettingsUtils.resolveIntent(option, context.packageName)
        ?: throw SettingsError.Unavailable(
          "Requested setting is not available on Android",
        )

    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

    if (intent.resolveActivity(context.packageManager) == null) {
      throw SettingsError.Unavailable(
        "No activity found to handle settings intent",
      )
    }

    return intent
  }
}

File: packages/settings/ios/Sources/SettingsPlugin/SettingsImpl.swift
import Foundation
import UIKit

/**
 Native iOS implementation for the Settings plugin.

 Responsibilities:
 - Perform platform-specific operations
 - Throw typed SettingsError values on failure

 Forbidden:
 - Accessing CAPPluginCall
 - Referencing Capacitor APIs
 - Constructing JS payloads
 */
@objc public final class SettingsImpl: NSObject {

    // Properties
    private var config: SettingsConfig?

    // Initializer
    override init() {
        super.init()
    }

    // MARK: - Configuration

    /**
     Applies static plugin configuration.

     This method must be called exactly once
     during plugin initialization.
     */
    func applyConfig(_ config: SettingsConfig) {
        precondition(
            self.config == nil,
            "SettingsImpl.applyConfig(_:) must be called exactly once"
        )
        self.config = config
        SettingsLogger.verbose = config.verboseLogging

        SettingsLogger.debug(
            "Configuration applied. Verbose logging:",
            config.verboseLogging
        )
    }

    // MARK: - Settings

    /**
     Opens the requested iOS settings page.

     - Parameter option: JS-facing settings key
     - Throws: SettingsError if the operation cannot be completed
     */
    func open(option: String?) throws {
        guard let url = SettingsUtils.resolveSettingsURL(for: option) else {
            throw SettingsError.unavailable(
                "Requested setting is not available on iOS"
            )
        }

        guard UIApplication.shared.canOpenURL(url) else {
            throw SettingsError.unavailable(
                "Cannot open settings URL"
            )
        }

        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:])
        }
    }
}
